scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala

```diff
@@ -0,0 +1,157 @@
+  * a companion class initialization lock.
+  */
+class CompanionRaceTest() {
+
+  /**
+    * Make a separate class loader for a test, and load the [[TestBridge]] class.
+    * The bug occurs while loading classes, hence the need for a class loader per test.
+    * @return The TestBridge class in the new class loader.
+    */
+  def createForeignClazz(): Class[_] = {
+    val systemClassLoader = this.getClass.getClassLoader()
+
+    val classLoader = new ClassLoader(systemClassLoader.getParent) {
+      override def findClass(name: String): Class[_] = {
+        val path = name.replace('.', '/').concat(".class")
+        val resOpt = systemClassLoader.getResourceAsStream(path)
+        Option(resOpt) match {
+          case Some(res) =>
+            try {
+              val bytes = Stream.continually(res.read).takeWhile(_ != -1).map(_.toByte).toArray
+              defineClass(name, ByteBuffer.wrap(bytes), null)
+            } catch {
+              case e: IOException => throw new ClassNotFoundException(name, e)
+            }
+          case None => throw new ClassNotFoundException(name)
+        }
+      }
+      override def loadClass(name: String, resolve: Boolean) =
+        super.loadClass(name, true)
+    }
+    classLoader.loadClass(new TestBridge().getClass.getName)
+  }
+
+  class TestSetup() extends Race {
+
+    private val bridgeClazz = createForeignClazz()
+    def method(name: String) = bridgeClazz.getMethod(name)
+    private var foreignObject = bridgeClazz.getConstructor().newInstance()
+
+    def changeInstance(): Unit = {
+      foreignObject = bridgeClazz.getConstructor().newInstance()
+    }
+
+    def value(): String = s"""${method("value").invoke(foreignObject)}"""
+    def withName(): String = s"""${method("withName").invoke(foreignObject)}"""
+
+  }
+  @Test
+  def valueBeforeRace(): Unit = {
+    val testSetup = new TestSetup()
+    testSetup.value()
+    testSetup.changeInstance()
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+  @Test
+  def raceToDeadlock(): Unit = {
+    val testSetup = new TestSetup()
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+
+  /**
+    * This is [[CompanionRaceTest.raceToDeadlock]] modified to avoid the race
+    * by running one of the halves first, which initializes all the classes,
+    * we then change the instance in `testSetup` and run the race, which does
+    * not deadlock - showing that the class initialization lock is involved in
+    * the deadlock.
+    */
+  @Test
+  def withNameBeforeRace(): Unit = {
+    val testSetup = new TestSetup()
+    testSetup.withName()
+    testSetup.changeInstance()
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+
+}
+
+abstract class TestCompanion(properties: Option[String]) {}
+
+object TestCompanion {
+  // Set this to false to exhibit the deadlock.
+  private val WORKAROUND = true
+  lazy val symbols: Set[TestCompanion] = {
+    Thread.sleep(10)
```

Comment:
Ha! At what time threshold do you start to see this behavior reliably? I assume it's rare without the sleep...

Reply:
I don't recall why there is this sleep.

Reply:
testing again today the sleep seems unnecessary; I have removed it